File: October_2025/potd_07_10_2025.py
# POTD: Sort Colors
# URL: https://practice.geeksforgeeks.org/problems/sort-colors
# Github username: <BhuwanSKumar>

def sort_colors(nums):
    """
    Sort an array containing 0s, 1s, and 2s (representing colors) in-place.
    Uses Dutch National Flag algorithm.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    low, mid, high = 0, 0, len(nums) - 1

    while mid <= high:
        if nums[mid] == 0:
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif nums[mid] == 1:
            mid += 1
        else:  # nums[mid] == 2
            nums[mid], nums[high] = nums[high], nums[mid]
            high -= 1
    return nums

# Example Tests
if __name__ == "__main__":
    assert sort_colors([2,0,2,1,1,0]) == [0,0,1,1,2,2]
    assert sort_colors([]) == []
    assert sort_colors([0,0,0]) == [0,0,0]
    print("All test cases passed!")

File: October_2025/potd_07_10_2025.java

// POTD: Sort Colors
// URL: https://practice.geeksforgeeks.org/problems/sort-colors
// Github username: <BhuwanSKumar>
import java.util.Arrays;

public class potd_07_10_2025 {

    public static void sortColors(int[] nums) {
        int low = 0, mid = 0, high = nums.length - 1;

        while (mid <= high) {
            switch (nums[mid]) {
                case 0:
                    int temp = nums[low];
                    nums[low] = nums[mid];
                    nums[mid] = temp;
                    low++;
                    mid++;
                    break;
                case 1:
                    mid++;
                    break;
                case 2:
                    int temp2 = nums[mid];
                    nums[mid] = nums[high];
                    nums[high] = temp2;
                    high--;
                    break;
            }
        }
    }

    public static void main(String[] args) {
        int[] arr1 = { 2, 0, 2, 1, 1, 0 };
        sortColors(arr1);
        System.out.println(Arrays.toString(arr1));

        int[] arr2 = {};
        sortColors(arr2);
        System.out.println(Arrays.toString(arr2));

        int[] arr3 = { 0, 0, 0 };
        sortColors(arr3);
        System.out.println(Arrays.toString(arr3));
    }
}